File: src/main/java/com/gitee/planners/module/script/SingletonScript.java
package com.gitee.planners.module.script;

import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 单例脚本封装
 * <p>
 * 替代 SingletonFluxonScript，封装单个脚本源码并提供执行入口。
 */
public class SingletonScript {

    private static final Pattern NESTED_PATTERN = Pattern.compile("\\{\\{(.+?)}}");

    private final String action;

    public SingletonScript(String source) {
        this.action = source != null ? source : "";
    }

    public boolean isNotNull() {
        return !action.isEmpty();
    }

    public String getAction() {
        return action;
    }

    /**
     * 执行脚本，根据 async 标志决定同步/异步
     */
    public CompletableFuture<Object> run(ScriptOptions options) {
        if (action.isEmpty()) {
            return CompletableFuture.completedFuture(null);
        }
        if (options.isAsync()) {
            return CompletableFuture.supplyAsync(() -> ScriptManager.eval(action, options));
        }
        return CompletableFuture.completedFuture(ScriptManager.eval(action, options));
    }

    /**
     * 同步执行
     */
    public Object eval(ScriptOptions options) {
        if (action.isEmpty()) return null;
        return ScriptManager.eval(action, options);
    }

    /**
     * 同步执行 (无上下文)
     */
    public Object eval() {
        return eval(new ScriptOptions());
    }

    /**
     * 替换文本中的嵌套脚本表达式 {{expr}}
     */
    public static String replaceNested(String text, ScriptOptions options) {
        Matcher matcher = NESTED_PATTERN.matcher(text);
        StringBuffer sb = new StringBuffer();
        while (matcher.find()) {
            String expr = matcher.group(1);
            Object result = new SingletonScript(expr).eval(options);
            matcher.appendReplacement(sb, Matcher.quoteReplacement(result != null ? result.toString() : ""));
        }
        matcher.appendTail(sb);
        return sb.toString();
    }
}

File: src/main/kotlin/com/gitee/planners/module/script/SingletonScript.kt
package com.gitee.planners.module.script

import java.util.concurrent.CompletableFuture

/** 启动期预编译、运行期只接受显式参数的 Nova 纯函数。 */
open class SingletonScript(
    source: String,
    sourceId: String,
    parameters: List<String> = emptyList()
) {

    val action: String = source
    private val unit: NovaScriptUnit?

    init {
        if (action.isEmpty()) {
            unit = null
        } else {
            unit = ScriptManager.compileExpression(sourceId, parameters, action)
        }
    }

    /** @return 当前表达式是否包含可执行内容。 */
    fun isNotNull(): Boolean {
        return action.isNotEmpty()
    }

    /** 按调用方指定的提交方式执行纯函数。 */
    fun run(async: Boolean, vararg arguments: Any?): CompletableFuture<Any?> {
        if (async) {
            return CompletableFuture.supplyAsync {
                eval(*arguments)
            }
        }
        return CompletableFuture.completedFuture(eval(*arguments))
    }

    /** 同步执行已编译纯函数。 */
    fun eval(vararg arguments: Any?): Any? {
        val compiled = unit
        if (compiled == null) {
            return null
        }
        return ScriptManager.invokePure(compiled, *arguments)
    }
}
